lib/pymedphys/_streamlit/apps/wlutz/_dbf.py

```diff
@@ -0,0 +1,136 @@
+from pymedphys._imports import streamlit as st
+
+from pymedphys._streamlit.utilities import dbf
+
+
+DBF_DATABASE_LOADING_CONFIG = {
+    "frame": {
+        "filename": "FRAME.dbf",
+        "columns_to_keep": ["DBID", "PIMG_DBID", "DELTA_MS"],
+        "column_rename_map": {"DBID": "FRAME_DBID"},
+    },
+    "patimg": {
+        "filename": "PATIMG.dbf",
+        "columns_to_keep": [
+            "DBID",
+            "DICOM_UID",
+            "IMG_DATE",
+            "IMG_TIME",
+            "PORT_DBID",
+            "ORG_DTL",
+        ],
+        "column_rename_map": {"DBID": "PIMG_DBID", "ORG_DTL": "machine_id"},
+    },
+    "port": {
+        "filename": "PORT.dbf",
+        "columns_to_keep": ["DBID", "TRT_DBID", "ID"],
+        "column_rename_map": {"DBID": "PORT_DBID", "ID": "port"},
+    },
+    "trtmnt": {
+        "filename": "TRTMNT.dbf",
+        "columns_to_keep": ["DBID", "PAT_DBID", "ID"],
+        "column_rename_map": {"DBID": "TRT_DBID", "ID": "treatment"},
+    },
+    "patient": {
+        "filename": "PATIENT.dbf",
+        "columns_to_keep": ["DBID", "ID", "LAST_NAME", "FIRST_NAME"],
+        "column_rename_map": {"DBID": "PAT_DBID", "ID": "patient_id"},
+    },
+}
+
+
+def loading_and_merging_dbfs(database_directory, refresh_cache):
+    patimg = load_dbf(database_directory, refresh_cache, "patimg")
+
+    dates = pd.to_datetime(patimg["IMG_DATE"], format="%Y%m%d").dt.date
+    date_options = dates.sort_values(ascending=False).unique()
+
+    selected_date = st.selectbox("Date", options=date_options)
+    patimg_filtered_by_date = patimg.loc[dates == selected_date]
+
+    merged = patimg_filtered_by_date
+
+    for database_key, merge_key in [
+        ("port", "PORT_DBID"),
+        ("trtmnt", "TRT_DBID"),
+        ("patient", "PAT_DBID"),
+    ]:
+        dbf_to_be_merged = load_dbf(database_directory, refresh_cache, database_key)
+        merged = merged.merge(dbf_to_be_merged, left_on=merge_key, right_on=merge_key)
+
+    timestamps_string = (
+        merged["IMG_DATE"].astype("str")
+        + "T"
+        + merged["IMG_TIME"].astype("str")
+        + "000"
+    )
+
+    merged["datetime"] = pd.to_datetime(timestamps_string, format="%Y%m%dT%H%M%S%f")
+    merged["time"] = merged["datetime"].dt.time
+
+    merged = merged[
+        [
+            "time",
+            "machine_id",
+            "patient_id",
+            "treatment",
+            "port",
+            "datetime",
+            "LAST_NAME",
+            "FIRST_NAME",
+            "PIMG_DBID",
+            "DICOM_UID",
+        ]
+    ]
+
+    return merged
+
+
+def dbf_to_pandas_without_cache(path):
+    return pd.DataFrame(iter(dbf.get_dbf_table(path)))
+
+
+@st.cache()
+def dbf_to_pandas_with_cache(path):
+    return [dbf_to_pandas_without_cache(path)]
+
+
+def dbf_to_pandas(path, refresh_cache=False):
+    result = dbf_to_pandas_with_cache(path)
+
+    if refresh_cache:
+        result[0] = dbf_to_pandas_without_cache(path)
+
+    return result[0]
```

Comment:
I assume the use of the list here is to make it compatible with st.cache? This is a little confusing first time looking at it but makes sense if that an st.cache requirement.

Reply:
Yeah, it is confusing. I should think of a clearer way to do it. Essentially I wanted to create a mutable output so that I can mutate the cache. This provides the functionality for the user to "re-read" the database. When a streamlit function is cached it only runs once and then stores the results. This is an issue though if the database updates. I may just have to remove the cache altogether in the end, or cache a different part of the pipeline instead. Having the database not update and be cached may just be a bad idea in and of itself.

I'll leave this here for now, but will certainly consider removing the cache altogether as I iterate.

Reply:
premature optimisation is not good.
caching something that you know a-priori is likely to change "underneath" and with no visible warning is a bad idea.
Is the performance hit of going to the db that horrific?
If so, and if you think that db changes that are relevant to the running application could happen with some frequency, reconsider the architecture.  "Stuff changes frequently" tends to motivate an event driven architecture.

Reply:
During my usage this was live running on one monitor while I was coding on the other. It certainly wasn't premature as it was slowing down my ability to see results down the line. I do agree however that caching this one is a bad idea. And it is likely worth considering having a watchdog based listener trigger a rerun in this case.

For now, the caching is helping me iterate on code down the line. So for now I'll leave it as is.

Reply:
> premature optimisation is not good.
> caching something that you know a-priori is likely to change "underneath" and with no visible warning is a bad idea.
> Is the performance hit of going to the db that horrific?

To explain myself here a bit better. The database is actually at a separate site, through an SSH tunnel. It's not that it's changing regularly, in fact when a previous date is selected for data viewing, it is not going to be changing at all. For now, so that I didn't have to wait on the internet handshake each time I changed a line of code I found caching this read was worthwhile.